hwloc autoadjust

psychocrypt opened the pull request:
- `CMakeLists.txt` add find code to search for `hwloc`
- `autoAdjustHwloc.hpp` search optimal config based on hwloc topology
- update `README.txt`

If hwloc is disabled with the cmake flag `-DHWLOC_ENABLE=OFF` than the old auto adjust implementation is used.
 
## Tests

-  [x] linux with L3 cache
  - AMD Opteron 62766 4 Socket
  - Intel i7 4600U
  - E5 2630 v3 2 Socket
  - ...
- [x] linux without L3 cache
- [x] windows (tested by @fireice-uk)

psychocrypt commented:
yes you are right libhwloc-dev. I will fix it.

fireice-uk commented:
Another minor issue, for odd numbered L3 caches it is better to over-load than under-load. So for example for my Linux laptop:

```
[2017-05-14 10:36:11] : Autoconf L3 size detected at 3072 KB.
[2017-05-14 10:36:11] : Autoconf core count detected as 4 on Linux.

**************** Copy&Paste ****************

"cpu_threads_conf" :
[
   { "low_power_mode" : false, "no_prefetch" : true, "affine_to_cpu" : 0 },
   { "low_power_mode" : false, "no_prefetch" : true, "affine_to_cpu" : 1 },
],
```

Whereas hwloc gives:

```
**************** Copy&Paste ****************

"cpu_threads_conf" :
[
   { "low_power_mode" : false, "no_prefetch" : true, "affine_to_cpu" : 0 },
],

**************** Copy&Paste ****************
```

psychocrypt commented:
But is this a general solution if we had a few byte more than a multple of 2MiB?
What is the rule to use more threads than cache available? Is it 1MiB or 512KiB?
How large is the difference in hash/s on your system?

fireice-uk commented:
On my laptop (I3-4010U) the difference is +30%
```
HASHRATE REPORT
| ID | 2.5s |  60s |  15m | ID | 2.5s |  60s |  15m |
|  0 | 16.4 | 16.6 | (na) |  1 | 30.5 | 29.4 | (na) |
-----------------------------------------------------
Totals:   47.0 46.0 (na) H/s
Highest:  47.1 H/s
````

Whereas single core is:
```
HASHRATE REPORT
| ID | 2.5s |  60s |  15m |
|  0 | 36.0 | 35.9 | (na) |
---------------------------
Totals:   36.0 35.9 (na) H/s
Highest:  36.1 H/s
```

So there is some cache interference (-15%), but the extra core running at half-speed makes up for it.

psychocrypt commented:
I will at a rule that if 1MiB is avail a new thread or two hashes per thread depending on the system will be used.

fireice-uk commented:
I will test the windows build soon. Are you looking to handle NUMA allocation internally or through the config?

fireice-uk commented:
By the way, are there any CPUs with fractional number of MB e.g. 2560 KB of cache?

psychocrypt commented:
You can download a precompiled hwloc static and dynamic lib for windows from the website.

The numa configuration is handled by the adjust class and mapped to the current way how threads are pinned.
The allocations are also fine because the OS is pinning memory to the numa node in the moment where it is touched the first time.
Never the less we can look if there is any improvement possible independent to this PR.

thx for testing the new feature.

psychocrypt commented:
> By the way, are there any CPUs with fractional number of MB e.g. 2560 KB of cache?

Not sure maybe old CPUs or atom.

fireice-uk commented:
Hold on, unless I'm wildly wrong we don't have any code to actually allocate the node right now? 

It will surely need to be something similar to thd_setaffinity in the current code?

psychocrypt commented:
Maybe I understand you last post wrong.

IMO there is no config needed for memory.
A thread is pinned by setadfinity to a core of a NUMA node. This is everything we must do to take care that the cache is reused and not flushed during a hashround and between (avoid TLB flush). The memory is allocated on the main memory near to the thread on the first touch (not at the line where it is allocated). 

My implementation with hwloc handle each socket and package within a system independently so that we should get always the optimal mapping.

fireice-uk commented:
I never worked with NUMA so I might be wrong, feel free to talk me through how those steps are handled:

1. Thread is created. At this point OS will allocate stack memory.
2. We set the affinity. At this point OS knows where we want to run.
3. We allocate the 2MB buffer on the heap using the standard functions.

How can OS know where we want to have our memory in 3 or 1?

psychocrypt commented:
For 1. I am not sure but for 3. the OS knows that your thread is pinned and selected the memory near ro the core (NUMA node).
I need to check if it is also true for windows but this is the way like linux is doing it.
I will open an issue today and point to this discussion, this allow to focus to this issue later again.
But also if we get memory on the wrong numa note this should not be an issue for the hash algo because the scratchpad is already in the cache if we solve one hash and get work for the next.

fireice-uk commented:
It won't always be in the cache, there are occasional misses even with full L3 allocation. One of the points I noted when testing on multi-cpu, multi-ram systems (on Linux) is that the results varied by around 5% but were constant during the run, and usually during a run shortly after, whereas they were most likely to change after reboot.

psychocrypt commented:
Tested on a system without L3 cache: `Intel(R) Core(TM)2 Duo CPU     T6570  @ 2.10GHz
`
I added the info that the config is not optimal because on the T6570 the suggestion is 1 thread (because the cpu has 2MiB RAM) but I get the best performance with two threads. Maybe it is coming from the fact that this CPU has no AES-ni support.

I fixed the issue in the README.

fireice-uk commented:
Sorry for the delay with Windows testing. Jobs are starting to clear up so I finally had some time to clear the backlog. Here is the windows output:

```
The configuration for 'cpu_threads_conf' in your config file is 'null'.
The miner evaluates your system and prints a suggestion for the section `cpu_thr
eads_conf` to the terminal.
The values are not optimal, please try to tweak the values based on notes in con
fig.txt.
Please copy & paste the block within the asterisks to your config.


**************** Copy&Paste ****************

"cpu_threads_conf" :
[
   { "low_power_mode" : false, "no_prefetch" : true, "affine_to_cpu" : 0 },
   { "low_power_mode" : false, "no_prefetch" : true, "affine_to_cpu" : 2 },
],

**************** Copy&Paste ****************
Press any key to exit.
```

Which seems correct. There is a two-line modification that needs to be made to smooth over the differences between gcc and msvc (as always), but that's nothing major.

I elected to link hwloc as a static lib so final exe size will go up 100KB or so.

fireice-uk commented:
This is hwloc-ls output on Windows (same machine):

```
+----------------------------------------+
¦ Machine (3167MB total)                 ¦
¦                                        ¦
¦ +------------------------------------+ ¦
¦ ¦ +--------------------------------+ ¦ ¦
¦ ¦ ¦ NUMANode P#0 (3167MB)          ¦ ¦ ¦
¦ ¦ +--------------------------------+ ¦ ¦
¦ ¦                                    ¦ ¦
¦ ¦ +--------------------------------+ ¦ ¦
¦ ¦ ¦ Package P#0                    ¦ ¦ ¦
¦ ¦ ¦                                ¦ ¦ ¦
¦ ¦ ¦ +----------------------------+ ¦ ¦ ¦
¦ ¦ ¦ ¦ L3 (3072KB)                ¦ ¦ ¦ ¦
¦ ¦ ¦ +----------------------------+ ¦ ¦ ¦
¦ ¦ ¦                                ¦ ¦ ¦
¦ ¦ ¦ +------------+  +------------+ ¦ ¦ ¦
¦ ¦ ¦ ¦ L2 (256KB) ¦  ¦ L2 (256KB) ¦ ¦ ¦ ¦
¦ ¦ ¦ +------------+  +------------+ ¦ ¦ ¦
¦ ¦ ¦                                ¦ ¦ ¦
¦ ¦ ¦ +------------+  +------------+ ¦ ¦ ¦
¦ ¦ ¦ ¦ L1d (32KB) ¦  ¦ L1d (32KB) ¦ ¦ ¦ ¦
¦ ¦ ¦ +------------+  +------------+ ¦ ¦ ¦
¦ ¦ ¦                                ¦ ¦ ¦
¦ ¦ ¦ +------------+  +------------+ ¦ ¦ ¦
¦ ¦ ¦ ¦ L1i (32KB) ¦  ¦ L1i (32KB) ¦ ¦ ¦ ¦
¦ ¦ ¦ +------------+  +------------+ ¦ ¦ ¦
¦ ¦ ¦                                ¦ ¦ ¦
¦ ¦ ¦ +------------+  +------------+ ¦ ¦ ¦
¦ ¦ ¦ ¦ Core       ¦  ¦ Core       ¦ ¦ ¦ ¦
¦ ¦ ¦ ¦            ¦  ¦            ¦ ¦ ¦ ¦
¦ ¦ ¦ ¦ +--------+ ¦  ¦ +--------+ ¦ ¦ ¦ ¦
¦ ¦ ¦ ¦ ¦ PU P#0 ¦ ¦  ¦ ¦ PU P#2 ¦ ¦ ¦ ¦ ¦
¦ ¦ ¦ ¦ +--------+ ¦  ¦ +--------+ ¦ ¦ ¦ ¦
¦ ¦ ¦ ¦ +--------+ ¦  ¦ +--------+ ¦ ¦ ¦ ¦
¦ ¦ ¦ ¦ ¦ PU P#1 ¦ ¦  ¦ ¦ PU P#3 ¦ ¦ ¦ ¦ ¦
¦ ¦ ¦ ¦ +--------+ ¦  ¦ +--------+ ¦ ¦ ¦ ¦
¦ ¦ ¦ +------------+  +------------+ ¦ ¦ ¦
¦ ¦ +--------------------------------+ ¦ ¦
¦ +------------------------------------+ ¦
+----------------------------------------+
+----------------------------------------+
¦ Host: LAPTOP-440                       ¦
¦                                        ¦
¦ Indexes: physical                      ¦
¦                                        ¦
¦ Date: 21/05/2017 10:48:49              ¦
+----------------------------------------+
```

fireice-uk commented:
After some more research POPCNT is a fairly late addition to the x86 instruction set. I switched over to a calculated solution since there is no point in bothering with compatibility for something so small.

psychocrypt commented:
@fireice-uk Thanks for testing and the update. I will check your changes today night on my test system.

psychocrypt commented on autoAdjustHwloc.hpp:
```diff
@@ -0,0 +1,211 @@
+ */
+int compat_popcnt(size_t x) 
+{
+	x = x - ((x >> 1) & (size_t)~(size_t)0/3);
```
Instead of adding this little magic code we can also add a for loop what is human readable.
The code part is not time critical.
```C++
template< typename T >
int compat_popcnt( T x ) 
{
  size_t count = 0;
  for( int i=0; i < sizeof(T) - 1; ++i )
    if( ((x>>i) & 1) == 1 )
      ++count;
}
```

psychocrypt commented on autoAdjustHwloc.hpp:
```diff
@@ -0,0 +1,211 @@
+ * Portable Snippets - https://gitub.com/nemequ/portable-snippets
+ * Created by Evan Nemerson <evan@nemerson.com>
+ */
+int compat_popcnt(size_t x) 
```
We should change this to a member function

psychocrypt commented:
I verified the current changes on different systems.

fireice-uk commented:
Agreed. Can you push the change that you propose?

psychocrypt commented:
ready to merge, I update the PR

fireice-uk commented:
@psychocrypt Can we use this branch for NUMA allocation tests? I don't want to add a dependency solely to replicate existing code.

psychocrypt commented:
Do you mean the issue with the memory allocation.
This pull request is feature complete and I wouldn't add more changes to
this PR.


Am 24.05.2017 9:23 Vorm. schrieb "fireice-uk" <notifications@github.com>:

@psychocrypt <https://github.com/psychocrypt> Can we use this branch for
NUMA allocation tests? I don't want to add a dependency solely to replicate
existing code.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/fireice-uk/xmr-stak-cpu/pull/108#issuecomment-303640424>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AYsxtnU5X3xC9V6y9wQbu1qDQa4QnhMoks5r89sKgaJpZM4NaK3l>
.

fireice-uk commented:
Yes, I meant allocating memory proximal to the NUMA node. Ok. I will merge this one, but as I said - that's where the perf gain will come from. I don't mind having an extra dependency on the dev branch, but for release we should make sure that hwloc's presence is justified.
